### ISSUE TITLE
OSS-Fuzz 53330: heap-buffer-overflow in zeek::BuildOptionsVal

### ISSUE THREAD

**Issue by timwoj** [zeek/zeek#2609]
Fuzzer: zeek-packet-fuzzer
Detailed report: https://oss-fuzz.com/testcase-detail/5102080180486144
Minimized testcase: [clusterfuzz-testcase-minimized-zeek-packet-fuzzer-5102080180486144.zip](https://github.com/zeek/zeek/files/10077081/clusterfuzz-testcase-minimized-zeek-packet-fuzzer-5102080180486144.zip)

```
==742==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60c000bfedf9 at pc 0x7f02d903e9ee bp 0x7ffec35dabd0 sp 0x7ffec35dabc8
READ of size 1 at 0x60c000bfedf9 thread T0
    #0 0x7f02d903e9ed in zeek::BuildOptionsVal(unsigned char const*, int) zeek/src/IP.cc:28:22
    #1 0x7f02d90399c5 in zeek::IPv6_Hdr::ToVal(zeek::IntrusivePtr<zeek::VectorVal>) const zeek/src/IP.cc:209:19
    #2 0x7f02d9041ce1 in zeek::IPv6_Hdr::ToVal() const zeek/src/IP.cc:314:9
    #3 0x7f02d9041ce1 in zeek::IPv6_Hdr_Chain::ToVal() const zeek/src/IP.cc:710:22
    #4 0x7f02d9040221 in zeek::IP_Hdr::ToIPHdrVal() const zeek/src/IP.cc:356:44
    #5 0x7f02d9044793 in zeek::IP_Hdr::ToPktHdrVal(zeek::IntrusivePtr<zeek::RecordVal>, int) const zeek/src/IP.cc:377:31
    #6 0x7f02d9044209 in zeek::IP_Hdr::ToPktHdrVal() const zeek/src/IP.cc:365:9
    #7 0x7f02da2ebd1a in zeek::packet_analysis::IP::IPAnalyzer::AnalyzePacket(unsigned long, unsigned char const*, zeek::Packet*) zeek/src/packet_analysis/protocol/ip/IP.cc:238:59
    #8 0x7f02d9b9de93 in zeek::packet_analysis::Analyzer::ForwardPacket(unsigned long, unsigned char const*, zeek::Packet*, unsigned int) const zeek/src/packet_analysis/Analyzer.cc:120:25
    #9 0x7f02d9baf059 in zeek::packet_analysis::Manager::ProcessPacket(zeek::Packet*) zeek/src/packet_analysis/Manager.cc:131:17
    #10 0x572339 in LLVMFuzzerTestOneInput zeek/src/fuzzers/packet-fuzzer.cc:35:22
    #11 0x440423 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #12 0x42bb82 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:324:6
    #13 0x43142c in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:860:9
    #14 0x45a962 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #15 0x7f02d4d5e0b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/libc-start.c:308:16
    #16 0x421d4d in _start

0x60c000bfedf9 is located 0 bytes to the right of 121-byte region [0x60c000bfed80,0x60c000bfedf9)
allocated by thread T0 here:
    #0 0x56c50d in operator new[](unsigned long) /src/llvm-project/compiler-rt/lib/asan/asan_new_delete.cpp:98:3
    #1 0x7f02da471489 in std::__1::__unique_if<unsigned char []>::__unique_array_unknown_bound std::__1::make_unique<unsigned char []>(unsigned long) /usr/local/include/c++/v1/__memory/unique_ptr.h:733:28
    #2 0x7f02da471489 in zeek::detail::FuzzBuffer::Next() zeek/src/fuzzers/FuzzBuffer.cc:82:15
    #3 0x572193 in LLVMFuzzerTestOneInput zeek/src/fuzzers/packet-fuzzer.cc:23:19
    #4 0x440423 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #5 0x42bb82 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:324:6
    #6 0x43142c in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:860:9
    #7 0x45a962 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #8 0x7f02d4d5e0b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/libc-start.c:308:16
```

**Comment by timwoj** [comment]
Relevant comment from zeek-security issue:

> This one crashes handling a ip6_mob_hoti_type mobility option here:
> 
> off += sizeof(uint16_t) + sizeof(uint64_t);
> m->Assign(2, BuildOptionsVal(data + off, Length() - off));
> 
> Length() here returns 8. off is set to sizeof(ip6_mobility) which is 6, and then adds 2 and 8 to it. This means it passes a negative number as the length to BuildOptionsVal, which doesn't handle it well. All of the case statements in IPv6_Hdr::ToVal could use some length checking, or we could make BuildOptionsVal return a nullptr on an error and have the consumers check that before assigning.

**Comment by timwoj** [comment]
Fixed via d5a1eb162e11a890f59125ac3c2d87e0ebc2fcd9
